content/client-lib-development-guide/features.textile

```diff
@@ -485,29 +483,38 @@ h3(#realtime-connection). Connection
-** @(RTN16f)@ The @msgSerial@ component of the @recoveryKey@, unlike the other two components, is not sent to Ably, but rather is used to set the library internal @msgSerial@. (If the recover fails, the counter should be reset to 0 per "RTN15c3":#RTN15c3 )
+** @(RTN16i)@ Connection recovery is similar to connection resumption (see "RTN15c":#RTN15c), except that instead of the library resuming from a time at which that library instance was previously connected, it is doing so from external state provided in the client options, "(TO3i)":#TO3i. Since the library has no state at the time of connection, the channels must be explicitly attached by the user; continuity preservation is achieved by the @channelSerial@s for each channel being stored in the recovery key.
+** @(RTN16f)@ When a library is instantiated with the @recover@ client option, it should initialize its internal @msgSerial@ counter to the @msgSerial@ component of the @recoveryKey@. (If the recover fails, the counter should be reset to 0 per "RTN15c7":#RTN15c7 )
+** @(RTN16j)@ When a library is instantiated with the @recover@ client option, for every channel/channelSerial pair in the @recoveryKey@, it should instantiate a corresponding channel and set its "RTL15b":#RTL15b @channelSerial@.
+** @(RTN16k)@ When the library first connects to Ably after being instantiated with a @recover@ client option, it should add an additional @recover@ querystring param to the websocket request, set from the @connectionKey@ component of the @recoveryKey@. Once the library has successfully connected to Ably, it should never again supply a @recover@ querystring param.
+** @(RTN16g)@ @Connection#getRecoveryKey@ is function that returns a string which incorporates the @connectionKey@, the current @msgSerial@, and a collection of pairs of channel @name@ and current @channelSerial@ for every currently attached channel. This must be serialized in a way which is able to encode any unicode channel name. The SDK may assume that the recovery key will only be consumed by the same type of SDK, so this spec does not specify any particular serialization; however, the format should be forward-compatible through the same major version of the SDK.
+*** @(RTN16g1)@ SDKs which choose not to increment their public major version when implementing protocol v2.0 may continue to implement this as a @Connection#recoveryKey@ attribute until the next major version change, using a getter function that's backwards-compatible with attribute access where possible, else by recomputing the recovery key string on every message.
+** @(RTN16h)@ @Connection#getRecoveryKey@ should return @Null@ when a the SDK is in the @CLOSED@, @CLOSING@, @FAILED@, or @SUSPENDED@ states, or when it does not have a @connectionKey@ (for example, it has not yet become connected). The @Connection#key@ and @Connection#id@ should also be @Null@ at these times.
```

Comment:
Do we need to check if `connectionId` is null too? Currently I am only checking if `connectionKey` is null

Reply:
I am currently only checking 
```
                if (Key.IsEmpty() || InnerState.State == Realtime.ConnectionState.Closing
                                  || InnerState.State == Realtime.ConnectionState.Closed
                                  || InnerState.State == Realtime.ConnectionState.Failed
                                  || InnerState.State == Realtime.ConnectionState.Suspended)
                {
                    return null;
                }
```
While in cocoa, we also check if `id` should be empty

Reply:
That looks fine to me

Reply:
You mean I should also check for `id` ?

Reply:
no, what you have there already looks fine

Reply:
Sure, thanks : )

Reply:
@sacOO7 If this conversation needs to continue, then that will need to happen against the new pull request that @SimonWoolf will open against https://github.com/ably/specification/issues/8. Please resolve this conversation _here_ once you have re-raised your concern _there_, or if you are already happy that this matter is resolved. Thanks.